File: problems/sorting/bubble_sort.py
from typing import List

# for ascending order ----------
class Solution:
    def BubbleSort(self, nums: List[int], visualize: bool = False) -> List[int]:
        for i in range(len(nums)):
            for j in range(len(nums) - 1 - i):
                if visualize:
                    print(nums)
                    
                if nums[j] > nums[j + 1]:
                    nums[j], nums[j + 1] = nums[j + 1], nums[j]
        return nums

# for descending order ------------
class Solution:
    def BubbleSortReverse(self, nums: List[int], visualize: bool = False) -> List[int]:
        for i in range(len(nums)):
            for j in range(len(nums) - 1 - i):
                if visualize:
                    print(nums)
                
                if nums[j] < nums[j + 1]:
                    nums[j], nums[j + 1] = nums[j + 1], nums[j]
        return nums

# for both ascending & descending order --------------
class Solution:
    def BubbleSort(self, nums: List[int], reverse: bool = False, visualize: bool = False) -> List[int]:
        for i in range(len(nums)):
            for j in range(len(nums) - 1 - i):
                if visualize:
                    print(nums)
                
                if nums[j] < nums[j + 1] if reverse else nums[j] > nums[j + 1]:
                    nums[j], nums[j + 1] = nums[j + 1], nums[j]
        return nums


# Time: O(n^2)
# Space: O(n)

# obj = Solution()
# print(obj.BubbleSort([3,2,1,0,8,9,2])) 
                    
                
                

File: problems/sorting/bubble_sort.java
package sorting;

import java.util.*;

// for ascending order ------------------------------------
class Solution {
    public boolean visualize = false;

    public int[] BuubleSort(int[] arr) {
        for (int i = 0; i < arr.length; i++) {
            if (visualize) {
                System.out.println(Arrays.toString(arr));
            }

            for (int j = 0; j < arr.length - 1 - i; j++) {
                if (arr[j] > arr[j + 1]) {
                    int temp = arr[j];
                    arr[j] = arr[j + 1];
                    arr[j + 1] = temp;
                }
            }
        }
        return arr;
    }
}

// for descending order -----------------------------
class Solution2 {
    public boolean visualize = false;

    public int[] BuubleSortdes(int[] arr) {
        for (int i = 0; i < arr.length; i++) {
            if (visualize) {
                System.out.println(Arrays.toString(arr));
            }

            for (int j = 0; j < arr.length - 1 - i; j++) {
                if (arr[j] < arr[j + 1]) {
                    int temp = arr[j];
                    arr[j] = arr[j + 1];
                    arr[j + 1] = temp;
                }
            }
        }
        return arr;
    }
}

// for both ascending & descending order -----------------
class Solution3 {
    public boolean visualize = false;

    public int[] BubbleSort(int[] arr, boolean reverse) {
        for (int i = 0; i < arr.length; i++) {
            if (visualize) {
                System.out.println(Arrays.toString(arr));
            }

            for (int j = 0; j < arr.length - 1 - i; j++) {
                if (reverse ? (arr[j] < arr[j + 1]) : (arr[j] > arr[j + 1])) {
                    int temp = arr[j];
                    arr[j] = arr[j + 1];
                    arr[j + 1] = temp;
                }
            }
        }
        return arr;
    }
}

// public class bubble_sort {
//     public static void main(String[] args) {
//         Solution coolsol = new Solution();
//         coolsol.visualize = true;
//         System.out.println(Arrays.toString(coolsol.BuubleSort(new int[]{4, 5, 1, 2, 6, 9, 4, 1})));
//     }
// }